drivers/sdhc/sdhci_common.c

```diff
@@ -0,0 +1,794 @@
+			       bool read)
+{
+	uint32_t *offs = (uint32_t *)data;
+	int i;
+
+	for (i = 0; i < size; i += 4, offs++) {
+		if (read) {
+			*offs = sys_read32(sdhci_ctx->reg_base + SDHCI_BUFFER);
+		} else {
+			sys_write32(*offs, sdhci_ctx->reg_base + SDHCI_BUFFER);
+		}
+	}
+}
+
+static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+			       bool is_read)
+{
+	uint32_t block = data->blocks;
+	uint32_t reg_present_state;
+	uint32_t int_status, rdy_mask;
+	uint8_t *data_ptr = data->data;
+	int32_t timeout;
+
+	timeout = data->timeout_ms ? data->timeout_ms : 1;
+	timeout = timeout * 100;
+
+	if (is_read) {
+		rdy_mask = SDHCI_BUF_RD_ENABLE;
+	} else {
+		rdy_mask = SDHCI_BUF_WR_ENABLE;
+	}
+
+	while (timeout && block) {
+		int_status = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (int_status & SDHCI_INT_ERROR) {
+			LOG_ERR("sdhc:req: data xfer error int_status:%08x", int_status);
+			return -EIO;
+		}
+
+		reg_present_state = sys_read32(sdhci_ctx->reg_base + SDHCI_PRESENT_STATE);
+		if (reg_present_state & rdy_mask) {
+			sdhci_data_port_io(sdhci_ctx, data_ptr, data->block_size, is_read);
+			data_ptr += data->block_size;
+			block--;
+		} else {
+			k_busy_wait(10);
+			timeout--;
+		}
+	};
+
+	if (!timeout) {
+		LOG_ERR("sdhc:req: transfer data timeout");
+		return -ETIMEDOUT;
+	}
+
+	return sdhci_poll_data_complete(sdhci_ctx, timeout / 100);
+}
+
+static void sdhci_prepare_dma(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+			      bool is_read)
+{
+	uintptr_t dma_addr;
+
+	sys_cache_data_flush_range(data->data, (data->blocks * data->block_size));
+
+	dma_addr = z_mem_phys_addr(data->data);
+	LOG_DBG("sdhc:req: SDHCI_DMA_ADDRESS %08lx", dma_addr);
+	sys_write32(dma_addr, sdhci_ctx->reg_base + SDHCI_DMA_ADDRESS);
+}
+
+static void sdhci_prepare_data(struct sdhci_common *sdhci_ctx, struct sdhc_command *cmd,
+			       struct sdhc_data *data, bool is_read)
+{
+	uint16_t transfer_mode = SDHCI_TRNS_BLK_CNT_EN;
+	uint32_t trans_bytes;
+
+	trans_bytes = data->blocks * data->block_size;
+	if (data->blocks > 1) {
+		transfer_mode |= SDHCI_TRNS_MULTI;
+	}
+
+	if (is_read) {
+		transfer_mode |= SDHCI_TRNS_READ;
+	}
+
+	if (sdhci_ctx->f_auto_cmd12 && sdhc_need_cmd12(cmd)) {
+		transfer_mode |= SDHCI_TRNS_ACMD12;
+	}
+
+	if (sdhci_ctx->f_use_dma) {
+		transfer_mode |= SDHCI_TRNS_DMA;
+		sdhci_prepare_dma(sdhci_ctx, data, is_read);
+	}
+
+	sys_write16(SDHCI_MAKE_BLKSZ(SDHCI_DEFAULT_BOUNDARY_ARG, data->block_size),
+		    sdhci_ctx->reg_base + SDHCI_BLOCK_SIZE);
+	sys_write16(data->blocks, sdhci_ctx->reg_base + SDHCI_BLOCK_COUNT);
+	sys_write16(transfer_mode, sdhci_ctx->reg_base + SDHCI_TRANSFER_MODE);
+}
+
+static void sdhci_cmd_done(struct sdhci_common *sdhci_ctx, struct sdhc_command *cmd,
+			   bool long_resp)
+{
+	cmd->response[0] = sys_read32(sdhci_ctx->reg_base + SDHCI_RESPONSE);
+	if (!long_resp) {
+		return;
+	}
+	cmd->response[1] = sys_read32(sdhci_ctx->reg_base + SDHCI_RESPONSE + 0x4);
```

Comment:
we can use `SDHCI_RESPONSE1`, `SDHCI_RESPONSE2` and `SDHCI_RESPONSE3` here

Reply:
yep.